src/quick-lint-js/fe/parse.cpp

```diff
@@ -36,833 +36,1023 @@
-    if (literal->is_null() || literal->is_regexp()) {
-      goto invalid;
-    }
-    break;
-  }
-
-  invalid:
-  default:
-    this->diag_reporter_->report(
-        diag_typescript_as_const_with_non_literal_typeable{
-            .expression = ast->span(),
-            .as_const = as_const_span,
-        });
-    break;
+  void parser::error_on_invalid_as_const(expression *ast,
+                                         source_code_span as_const_span)
+  {
+    ast = ast->without_paren();
+    switch (ast->kind())
+    {
+    case expression_kind::dot:
+    case expression_kind::array:
+    case expression_kind::object:
+      break;
+
+    case expression_kind::literal:
+    {
+      auto *literal = static_cast<expression::literal *>(ast);
+      if (literal->is_null() || literal->is_regexp())
+      {
+        goto invalid;
+      }
+      break;
+    }
+
+    invalid:
+    default:
+      this->diag_reporter_->report(
+          diag_typescript_as_const_with_non_literal_typeable{
+              .expression = ast->span(),
+              .as_const = as_const_span,
+          });
+      break;
+    }
   }
-}
 
-void parser::error_on_pointless_compare_against_literal(
-    expression::binary_operator* ast) {
-  auto is_comparison_operator = [](string8_view s) -> bool {
-    return s == u8"=="_sv || s == u8"==="_sv || s == u8"!="_sv ||
-           s == u8"!=="_sv;
-  };
+  void parser::error_on_pointless_compare_against_literal(
+      expression::binary_operator *ast)
+  {
+    auto is_comparison_operator = [](string8_view s) -> bool
+    {
+      return s == u8"=="_sv || s == u8"==="_sv || s == u8"!="_sv ||
+             s == u8"!=="_sv;
+    };
+
+    for (span_size i = 0; i < ast->child_count() - 1; i++)
+    {
+      source_code_span op_span = ast->operator_spans_[i];
+      if (is_comparison_operator(op_span.string_view()))
+      {
+        this->check_compare_against_literal(ast->child(i), ast->child(i + 1),
+                                            op_span);
+      }
+    }
+  }
 
-  for (span_size i = 0; i < ast->child_count() - 1; i++) {
-    source_code_span op_span = ast->operator_spans_[i];
-    if (is_comparison_operator(op_span.string_view())) {
-      this->check_compare_against_literal(ast->child(i), ast->child(i + 1),
-                                          op_span);
+  void parser::warn_on_mistyped_strict_inequality_operator(source_code_span op_span)
+  {
+    // op span is "bang_span" (?)
+
+    // How do I go about doing this?
+    // supposedly this is to the left of a bang, I think...
+
+    // Pseudo code due to knowlege gaps:
+
+    // check if lhs is enclosed within ()
+    // if (lhs is enclosed within parentheses) {
+    //   then null assertion was intended.
+    //   return
+    // }
+    //
+    // else, then check if rhs is "==" operator (equals)
+    // if (rhs is '==' ) {
+    // then report error
+
+    // print(op_span.string_view()); **CANT USE** (how do I know what op_span is??)
```

Comment:
op_span is a source_code_span object. We don't have code for printing a source_code_span... (I normally use gdb to debug this kind of thing, so I haven't added printing support.)

I added support in commit 518915e2b1b13d9ec1945c0d2003c825ff728779. You can apply that patch by running the following Git commands:

```bash
# In the following command, 'upstream' is your name
# for the https://github.com/quick-lint/quick-lint-js/ fork.
git fetch upstream
git cherry-pick 518915e2b1b13d9ec1945c0d2003c825ff728779
```

Then you can write this if you want to printf debug:
```
std::cout << op_span << '\n';
```

Reply:
> I added support in commit [518915e](https://github.com/quick-lint/quick-lint-js/commit/518915e2b1b13d9ec1945c0d2003c825ff728779). You can apply that patch by running the following Git commands:
> 
> ```shell
> # In the following command, 'upstream' is your name
> # for the https://github.com/quick-lint/quick-lint-js/ fork.
> git fetch upstream
> git cherry-pick 518915e2b1b13d9ec1945c0d2003c825ff728779
> ```

Due to Key issues, I can only use the desktop client for GitHub, Do you know how I could apply that patch on the Desktop client?